src/cliffordsynthesis/encoding/STQGatesEncoder.cpp

```diff
@@ -0,0 +1,236 @@
+//
+// Created by Velsh Aleksei on 16.06.23.
+//
+
+#include "cliffordsynthesis/encoding/STQGatesEncoder.hpp"
+
+#include "LogicTerm/LogicTerm.hpp"
+#include "utils/logging.hpp"
+
+namespace cs::encoding {
+
+using namespace logicbase;
+
+void encoding::STQGatesEncoder::createSingleQubitGateVariables() {
+  DEBUG() << "Creating single-qubit gate variables for the STQEncoding.";
+  vars.gS.reserve(T);
+  for (std::size_t t = 0U; t < T; ++t) {
+    auto& timeStep = vars.gS.emplace_back();
+    timeStep.reserve(SINGLE_QUBIT_GATES_FOR_STQ_ENCODING.size());
+    for (const auto gate : SINGLE_QUBIT_GATES_FOR_STQ_ENCODING) {
+      auto& g = timeStep.emplace_back();
+      g.reserve(N);
+      for (std::size_t q = 0U; q < N; ++q) {
+        const std::string gName = "g_" + std::to_string(t) + "_" +
+                                  toString(gate) + "_" + std::to_string(q);
+        TRACE() << "Creating variable " << gName;
+        g.emplace_back(lb->makeVariable(gName));
+      }
+    }
+  }
+}
+
+void encoding::STQGatesEncoder::createTwoQubitGateVariables() {
+  DEBUG() << "Creating two-qubit gate variables for the STQEncoding.";
+  vars.gC.reserve(T);
+  for (std::size_t t = 0U; t < T; ++t) {
+    auto& timeStep = vars.gC.emplace_back();
+    timeStep.reserve(N);
+    for (std::size_t ctrl = 0U; ctrl < N; ++ctrl) {
+      auto& control = timeStep.emplace_back();
+      control.reserve(N + 1U);
+      for (std::size_t trgt = 0U; trgt < N; ++trgt) {
+        const std::string gName = "g_" + std::to_string(t) + "_cx_" +
+                                  std::to_string(ctrl) + "_" +
+                                  std::to_string(trgt);
+        TRACE() << "Creating variable " << gName;
+        control.emplace_back(lb->makeVariable(gName));
+      }
+
+      // Add I to the two qubit gates
+      const std::string gName = "g_" + std::to_string(t) + "_" +
+                                toString(qc::OpType::None) + "_" +
+                                std::to_string(ctrl);
+      TRACE() << "Creating variable " << gName;
+      control.emplace_back(lb->makeVariable(gName));
+    }
+  }
+}
+
+void encoding::STQGatesEncoder::collectTwoQubitGateVariables(
+    const std::size_t pos, const std::size_t qubit, const bool target,
+    LogicVector& variables) const {
+  const auto& twoQubitGates = vars.gC[pos];
+  const auto  n             = twoQubitGates.size();
+  for (std::size_t q = 0; q < n; ++q) {
+    if (q == qubit) {
+      if (N > variables.size()) {
```

Comment:
I think I don't get this check. You could check for the `target` flag. This way you would also only add the variable for the identity once.